Decode SCALE encoded compact unsigned integers in Solidity

denalimarsh opened the pull request:
Implements Scale.sol, a contract that implements decoding mechanisms for SCALE encoded Substrate data.
- Decode compact unsigned integers according to the specification in https://github.com/centrifuge/go-substrate-rpc-client/blob/40b1b01ef45829256dfba2028e2e134dcf736f15/scale/codec.go.
- Currently successfully decodes all integers in the range [0, 1073741824], but have noticed slight discrepancies with decoding some integers 4294967296 and above.
- Run test with `truffle test test/test_scale.js`

**Gas Expenditure Report**
- Decoding [0, 63] average gas: 22,312
- Decoding [64, 16383] average gas: 22,550
- Decoding [16384, 1073741823] average gas: 22,918
- Decoding [1073741823, 4503599627370496] average gas: 25,737

philipstanislaus commented on ethereum/contracts/Scale.sol:
```diff
@@ -0,0 +1,86 @@
+        } else if(mode == 3) {
+            uint8 l = b >> 2;                         // remove mode bits
+            require(l <= 63,                          // Max upper bound of 536 is (67 - 4)
+                    "Not supported: l>63 encountered when decoding a compact-encoded uint");
```
The biggest number u256 can hold is 32 bytes. I’d propose to just drop support for bigger numbers here and throw if l > 32

denalimarsh replied:
That's definitely the easiest solution. Will dropping support for bigger numbers come back to bite us in the future?

Another potential option may be spreading the value across multiple memory slots i.e. 2+ uint256 variables.

philipstanislaus replied:
It will bite someone I guess, but every solution I can think of increases the gas cost for us, hence I would limit support for now.

The cleanest approach would likely be a big int type, which supports arithmetic operations, but that’s clearly out of scope.

denalimarsh replied:
By rejecting all compact uints that take up greater than 32 bytes, all numbers that hit case 3 will be rejected. The smallest possible case 3 number is 1073741824 (decoded), which previously passed the test but is now rejected.

Should we just remove the case 3 logic?

philipstanislaus replied:
Is the logic working now as expected? If it is still buggy, I would remove it. If it works, let's keep it in there.

denalimarsh replied:
Still buggy, removed in https://github.com/Snowfork/polkadot-ethereum/pull/51/commits/e9e7231e103fa771989a28187874db29300be165. We can bring it back and resolve the issues when applicable.

philipstanislaus commented on ethereum/contracts/Scale.sol:
```diff
@@ -0,0 +1,86 @@
+            uint8[] memory buf = new uint8[](l+4);
+            for(uint8 i = 1; i < l+4+1; i++) {        // read bytes from data[1:l+4]
+                uint8 bNext = readByteAtIndex(data, i);
+                buf[i-1] = bNext;
```
Instead of reversing below, we could also fill from back, right?

```Suggestion
                buf[l+4-i] = bNext;
```

denalimarsh replied:
Good idea, this will definitely save some gas at runtime.

denalimarsh replied:
If we remove the compact uint case 3 logic then both `uint8ArrayToUint256` and `reverse` can be removed. However, this suggestion still applies to the new method `reverseBytes`.

Before this PR is merged I'll either address it or make an issue for it.

denalimarsh replied:
Created https://github.com/Snowfork/polkadot-ethereum/issues/56.

philipstanislaus commented:
Thanks for your work on this 🙏

vgeddes reviewed (approved):
LGTM :+1:

We can hopefully use fixed encoding for u256 as discussed in the call today

denalimarsh commented:
Asking for another round of reviews due to new functionality.

I implemented uint256 decoding, added uint256 tests, and addressed revisions (see comments, have a couple of questions for @philipstanislaus).  @vgeddes could you plug some encoded uint256s generated by our substrate code into the `uint256` unit test and make sure it works for our use case?

vgeddes commented on ethereum/contracts/Scale.sol:
```diff
@@ -0,0 +1,124 @@
+    {
+        uint256 number;
+        for(uint i = 0; i < arr.length; i++){
+            number = number+uint(uint8(arr[i]))*(2**(8*(arr.length-(i+1))));
```
Could we have spaces between binary operators to make the code less dense and improve readability?

denalimarsh replied:
Fixed in https://github.com/Snowfork/polkadot-ethereum/commit/d3d5a2bc7a79b6949357ec8082aae20ce6b9c1b6

vgeddes commented on ethereum/contracts/Scale.sol:
```diff
@@ -0,0 +1,124 @@
+        uint8 b = readByteAtIndex(data, 0);           // read the first byte
+        uint8 mode = b & 3;                           // bitwise operation
+
+        if(mode == 0) {                               // [0, 63]
```
According to the recommended style guide at https://solidity.readthedocs.io/en/v0.5.3/style-guide.html there should be a space between the `if` and the first parenthesis.

denalimarsh replied:
Noted, fixed in https://github.com/Snowfork/polkadot-ethereum/pull/51/commits/d3d5a2bc7a79b6949357ec8082aae20ce6b9c1b6

vgeddes commented on ethereum/contracts/Scale.sol:
```diff
@@ -0,0 +1,124 @@
+
+        if(mode == 0) {                               // [0, 63]
+            return b >> 2;                            // right shift to remove mode bits
+        } else if(mode == 1) {                        // [64, 16383]
```
same here

denalimarsh replied:
https://github.com/Snowfork/polkadot-ethereum/pull/51/commits/d3d5a2bc7a79b6949357ec8082aae20ce6b9c1b6

vgeddes commented on ethereum/contracts/Scale.sol:
```diff
@@ -0,0 +1,124 @@
+            r <<= 6;                                  // multiply by * 2^6
+            r += b >> 2;                              // right shift to remove mode bits
+            return r;
+        } else if(mode == 2) {                        // [16384, 1073741823]
```
same here

denalimarsh replied:
https://github.com/Snowfork/polkadot-ethereum/pull/51/commits/d3d5a2bc7a79b6949357ec8082aae20ce6b9c1b6

philipstanislaus commented on ethereum/contracts/Scale.sol:
```diff
@@ -40,6 +40,44 @@ contract Scale {
+        returns (uint256)
+    {
+        uint256 number;
+        for(uint i = 0; i < arr.length; i++){
```
Again, to save gas, we could just iterate backwards over the array instead of reversing it, right?

denalimarsh replied:
Created https://github.com/Snowfork/polkadot-ethereum/issues/56 for this. Small, but want to get this milestone functional and submitted before addressing some of these smaller optimizations.

philipstanislaus reviewed (changes requested):
Nice, thanks!

denalimarsh commented on ethereum/package.json:
```diff
@@ -1,14 +1,17 @@
     "web3-utils": "^1.2.9"
+  },
+  "devDependencies": {
+    "@polkadot/types": "^1.28.1"
```
Removed in https://github.com/Snowfork/polkadot-ethereum/pull/51/commits/42ee4943545e90ff2c99159e5140d4cbc5cbe716.
